bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v2/PageImpl.java

```diff
@@ -311,4 +326,42 @@ public String getMainContentSelector() {
         }
         return htmlPageItems;
     }
+
+    @Override
+    public @Nullable String getCanonicalLink() {
+        try {
+            SeoTags seoTags = resource.adaptTo(SeoTags.class);
+            return seoTags != null ? seoTags.getCanonicalUrl() : null;
+        } catch (NoClassDefFoundError ex) {
+            return pathProcessor.externalize(currentPage.getPath(), request);
+        }
+    }
+
+    @Override
+    public @NotNull Map<Locale, String> getAlternateLanguageLinks() {
```

Comment:
Would be great, as general practice, if we lazy-cached the output of these methods so that we don't have to do all the processing on each call.

Reply:
It's not just the `adaptTo` part but also `SeoTags#getAleternateLanguages` and even the `Collections#unmodifiable*`

Reply:
If the `SeoTags#getAlternateLanguages()` is an expensive implementation worth caching, is an implementation detail and should be implemented by the implemented by the SeoTags implementation. The `PageImpl` as consumer should trust the implementation to perform optional. The `Collections#unmodifiableMap()` and `Collections#emptyMap()` should be inlined and so only in the former case we would safe a single object creation (a matter of nanoseconds in modern JMVs). 

However I don't have a strong opinion on this I will implement the change. In theory, it introduces complexity as you will have to test that subsequent calls return the same object.